### ISSUE TITLE
 Pandas Timestamp/Timedelta are not understood as subclasses of datetime.datetime and datetime.timedelta

### ISSUE THREAD

**Issue by Liam3851** [microsoft/python-type-stubs#110]
Originally reported as https://github.com/microsoft/pylance-release/issues/1986, referred to this repo.

## Environment data

-   Language Server version: 2021.10.2
-   OS and version: Windows 10 21H1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.11 Miniconda3
-   Pandas 1.3.3

## Expected behaviour

Pandas Timestamps and Timedeltas are subclasses of datetime.datetime and datetime.timedelta, respectively, with the latter showing up in the MRO of the former:

```python
In [1]: import pandas as pd

In [2]: pd.__version__
1.3.3

In [3]: pd.Timestamp.mro()
Out[3]:
[pandas._libs.tslibs.timestamps.Timestamp,
 pandas._libs.tslibs.timestamps._Timestamp,
 pandas._libs.tslibs.base.ABCTimestamp,
 datetime.datetime,
 datetime.date,
 object]

In [4]: pd.Timedelta.mro()
Out[4]:
[pandas._libs.tslibs.timedeltas.Timedelta,
 pandas._libs.tslibs.timedeltas._Timedelta,
 datetime.timedelta,
 object]

In [6]: issubclass(pd.Timestamp, datetime.datetime)
Out[6]: True

In [7]: issubclass(pd.Timestamp, datetime.date)
Out[7]: True

In [8]: issubclass(pd.Timedelta, datetime.timedelta)
Out[8]: True
```

## Actual behaviour

Pylance is not detecting this subclassing. For example this test code gives a type error in Pylance, saying "Argument of type "Timestamp" cannot be assigned to parameter "d" of type "date" in function "test".":
```python
import datetime
import pandas as pd

def test(d: datetime.date) -> datetime.date:
    return d

test(pd.Timestamp('2021-01-01'))
```

## Comments

On the other thread @jakebailey indicated that recording this subclassing correctly may have caused issues before and that @gramster would be aware of what they were. I'm not sure how reporting accurate subclass relationships could cause a bug in a typing system, though.

**Comment by Dr-Irv** [comment]
@gramster We may want to pull the current pandas source file `pandas/_libs/tslibs/timestamps.pyi` over to this project since it fixes this issue, and also addresses this oddity in type checking:
```python
from typing import List
import pandas as pd

tslist: List[pd.Timestamp] = list(pd.to_datetime(["2022-01-01", "2022-01-02"]))

sseries: pd.Series = pd.Series(tslist)


sseries_plus1 = sseries + pd.Timedelta(1, "d")

reveal_type(sseries_plus1)
```
produces
```
Type of "sseries_plus1" is "Timedelta"
```

But the type of `sseries_plus1` should be a `Series`.

This is due to two things:

1. In this project, `Series.__add__()` does not allow a `Timedelta` as an argument
2. In this project, `Timedelta.__radd__()` does not type the `other` argument, so it is getting used here to do the type resolution, producing the odd result.

If https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/timedeltas.pyi is copied over to this project, that should resolve this issue posted by @Liam3851 and handle (2).

But for (1), I'm not sure what to do.  We could add `Timedelta` as an arg to `Series.__add__()` (and other operators) but that's only valid if the corresponding series has a dtype of `Timestamp`, and I don't think the typing system can track that.

But I think you (@gramster ) are in a better place to make that decision than me doing a PR.

Maybe I should create a separate issue for (1) ??

**Comment by Dr-Irv** [comment]
> @Dr-Irv Wouldn't the objection in (1) apply to _any_ scalar? I mean you can't add 'a' to a float dtype Series either, but I don't think that currently gives an error, at least with pylance basic type checking on.

Yes, and since a `Series` can hold any type, knowing which types can be added to a series would be hard to handle via typing.

But I think the stub for `Series.__add__()` should be expanded to allow more types.

**Comment by Dr-Irv** [comment]
Original issue addressed in #156 but have to create test for comment at https://github.com/microsoft/python-type-stubs/issues/110#issuecomment-1007882648

**Comment by gramster** [comment]
We are no longer maintaining the pandas stubs; please open issues at https://github.com/pandas-dev/pandas-stubs
